### PR TITLE
Add common relationship methods to Model\ContactRelation

### PR DESCRIPTION
**Pull request by MrPetovan**
Part of #8918

- Introduce DBA::mergeConditions method
- Replace GContact relationship method contents with Model\ContactRelation method calls

### CODE REVIEW

**Comment by annando** [comment]
I had a look at the whole stuff as well and was unsure if this belongs to the `Contact` or `ContactRelation` class. We use the one table and return the other one - this had confused me :-)

Concerning the `listFollows`: I'm unsure here as well. Currently we do display all contact relations. If we only show one direction, I think we have to add a second link for the other direction.

This weekend I had worked on the whole `GContact` stuff as well and had removed all functions with the exception of the 4 "common contact" functions.

File: src/Database/DBA.php [hunk 1]
```diff
@@ -645,6 +645,34 @@ public static function collapseCondition(array $condition)
...
 	}
 
+	/**
+	 * Merges the provided conditions into a single collapsed one
+	 *
+	 * @param array ...$conditions One or more condition arrays
+	 * @return array A collapsed condition
+	 * @see DBA::collapseCondition() for the condition array formats
+	 */
+	public static function mergeConditions(array ...$conditions)
```
Comment (annando) [inline src/Database/DBA.php:655]: Have you checked this with the two possible ways of providing conditions?

File: src/Model/ContactRelation.php [hunk 2]
```diff
@@ -200,4 +206,333 @@ public static function isDiscoverable(string $url, array $contact = [])
...
+    WHERE `relation-cid` = ?
+      AND `follows`
+) 
+  AND `id` IN (
+    SELECT `cid`
+    FROM `contact-relation`
+    WHERE `relation-cid` = ?
+      AND `follows`
+)', $sourceId, $targetId]
+		);
```
Comment (annando) [inline src/Model/ContactRelation.php:491]: Some side note: I'm always confused when I see something split in so many rows. I totally lose context and have to collapse the stuff to understand whats happening here.

Also I'm confused about the indention with spaces here.

File: src/Database/DBA.php (hunk 1 above)
Comment (MrPetovan) [reply src/Database/DBA.php:655]: Yes, you can see in this function that the conditions are collapsed which normalize them from the two existing forms.

File: src/Model/ContactRelation.php (hunk 2 above)
Comment (MrPetovan) [reply src/Model/ContactRelation.php:491]: My IDE aligns the `AND` with the `WHERE` with spaces, can you please suggest an indentation that would be better readable for you?

**Comment by MrPetovan** [comment]
> Concerning the `listFollows`: I'm unsure here as well. Currently we do display all contact relations. If we only show one direction, I think we have to add a second link for the other direction.

My reasoning, confirmed with my girlfriend, is that we'd rather match interactions that both considered users willfully initiated. Since we do not intersect with the `user-contact` table, we can't hide the contacts who interacted with someone who blocked them, which we definitely do not want to show in the common contacts.

So now it shows both follows and public interactions, the same way we show follows by default on profile pages (except forums where it doesn't make sense).

File: src/Database/DBA.php (hunk 1 above)
Comment (annando) [reply src/Database/DBA.php:655]: I only had a short glimpse right now.

File: src/Model/ContactRelation.php (hunk 2 above)
Comment (annando) [reply src/Model/ContactRelation.php:491]: Your IDE is bad, take VSCode instead :grimacing: 

Concerning the indentation I guess most irritating is something like the single `)` in line 490. Also I still follow the (ancient) rule of keeping stuff into a single line that fit into a single line without scrolling, especially when it belongs together. (see my PR for this)

**Comment by annando** [comment]
I guess during the development phase we will have a look here and can adjust things if needed.

**Comment by MrPetovan** [comment]
Wait, should I move the whole `ContactRelation` class to`Contact\Relation` or just my changes?

**Comment by annando** [comment]
I'm unsure to move everything or not. I leave this decision to you, you are the expert.

**Comment by MrPetovan** [comment]
14 usages total, let's move it all.

**Comment by annando** [comment]
Okay. I wait until this is done.

**Comment by MrPetovan** [comment]
It is done.

**Comment by annando** [comment]
Conflict :grin:

**Comment by MrPetovan** [comment]
Fixed.

**Comment by MrPetovan** [comment]
I found a couple spots in GContact where I didn't add the necessary conditions.

**Comment by annando** [comment]
I'm waiting for the tests to complete.

**Pull request merged**